src/librustc_mir/interpret/memory.rs

```diff
@@ -879,4 +879,21 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> Memory<'a, 'mir, 'tcx, M> {
         }
         Ok(())
     }
+
+    pub fn force_ptr(
+        &self,
+        scalar: Scalar<M::PointerTag>,
+    ) -> InterpResult<'tcx, Pointer<M::PointerTag>> {
+        match scalar {
+            Scalar::Ptr(ptr) => Ok(ptr),
+            _ => M::int_to_ptr(scalar.to_usize(self)?, &self.extra)
+        }
+    }
+
+    pub fn force_bits(&self, scalar: Scalar<M::PointerTag>) -> InterpResult<'tcx, u128> {
```

Comment:
I know this PR isn't going to full intptrcast in one step, but I think it should still use `force_bits` $somewhere. Maybe start with just

* https://github.com/rust-lang/rust/blob/24ddd1615419be89828fb5628e3c14af86c08b01/src/librustc_mir/interpret/cast.rs#L251 (call `cast_from_int` after you get the bits via `force_bits`)
* https://github.com/rust-lang/rust/blob/24ddd1615419be89828fb5628e3c14af86c08b01/src/librustc_mir/interpret/place.rs#L472
* https://github.com/rust-lang/rust/blob/24ddd1615419be89828fb5628e3c14af86c08b01/src/librustc_mir/interpret/operator.rs#L313
* https://github.com/rust-lang/rust/blob/24ddd1615419be89828fb5628e3c14af86c08b01/src/librustc_mir/interpret/operator.rs#L350

Reply:
I'll do these changes and add a commit :P

Reply:
No please don't do anything on casts! That's just a red herring, people can still transmute stuff. We need to do conversion lazily, nor eagerly.

I thought we had discussed in detail where `force_ptr` is needed? Namely, the methods in "place.rs" and "operand.rs" that access memory. And then for `force_bits`, that would arise only inside Miri in the newest plan, in `ptr_op`.

Reply:
Oh wait, that cast stuff is for non-ptr-sized casts. Hm. Okay we should eventually `force_bits` there.

But that should be properly deduplicated with the int-to-int cass, so I don't think that method is the right place. Maybe let's keep that one for later.

Reply:
D: ok I haven't committed anything. What should I do next then?

Reply:
IOW, the run-tme dispatch on the value [here](https://github.com/rust-lang/rust/blob/24ddd1615419be89828fb5628e3c14af86c08b01/src/librustc_mir/interpret/cast.rs#L153) is just as ugly as the similar one in binops. We should dispatch only based on types, not on values... but we cannot yet due to pointers at integer types.

But what we can do is first dispatch based on whether the type is ptr-sized, and if yes proceed like now, and if not do basically `self.cast_from_int(self.force_bits(data))`.

Reply:
@christianpoveda what I [wrote here](https://github.com/rust-lang/rust/pull/61781#discussion_r293472671).

Reply:
Oh sorry my bad, this was only about `force_bits`. I should learn to read. :(

I agree with the following two:

https://github.com/rust-lang/rust/blob/24ddd1615419be89828fb5628e3c14af86c08b01/src/librustc_mir/interpret/place.rs#L472

https://github.com/rust-lang/rust/blob/24ddd1615419be89828fb5628e3c14af86c08b01/src/librustc_mir/interpret/operator.rs#L350

But [this one](https://github.com/rust-lang/rust/blob/24ddd1615419be89828fb5628e3c14af86c08b01/src/librustc_mir/interpret/operator.rs#L313) makes no sense to change as it already panics if there are no bits there.

Reply:
Ok, I've already substituted `to_ptr` by `force_ptr` in the last commit, if the `bin_op` change is going to happen in `miri`, are we done here?

Reply:
Oh also `force_bits` shouldn't assume pointer size, that makes it rather inconvenient to use. It should take the size as argument, like `to_bits` does.

But it can assert that if the size is NOT pointer size, it will never see a pointer value.

>  I've already substituted to_ptr by force_ptr in the last commit

I hope you mean "carefully replaced in a few places", not "substituted everywhere with `sed`". ;)

Reply:
Hahaha, yeah I've just replaced them on the places with the @RalfJung approval seal.

Reply:
> Oh sorry my bad, this was only about `force_bits`. I should learn to read. :(
> 
> I agree with the following two:
> 
> https://github.com/rust-lang/rust/blob/24ddd1615419be89828fb5628e3c14af86c08b01/src/librustc_mir/interpret/place.rs#L472
> 
> https://github.com/rust-lang/rust/blob/24ddd1615419be89828fb5628e3c14af86c08b01/src/librustc_mir/interpret/operator.rs#L350
> 
> But [this one](https://github.com/rust-lang/rust/blob/24ddd1615419be89828fb5628e3c14af86c08b01/src/librustc_mir/interpret/operator.rs#L313) makes no sense to change as it already panics if there are no bits there.

I'm going to do those changes then

Reply:
> Oh also `force_bits` shouldn't assume pointer size, that makes it rather inconvenient to use. It should take the size as argument, like `to_bits` does.
I'm going to do this change before the other ones then and will do a couple commits in a few minutes